pkg/providers/imagefamily/nodeimage.go

```diff
@@ -0,0 +1,132 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package imagefamily
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/karpenter-provider-azure/pkg/apis/v1alpha2"
+	"github.com/Azure/karpenter-provider-azure/pkg/operator/options"
+	"github.com/mitchellh/hashstructure/v2"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+)
+
+type NodeImage struct {
+	ID           string
+	Requirements scheduling.Requirements
+}
+
+type NodeImageProvider interface {
+	List(ctx context.Context, nodeClass *v1alpha2.AKSNodeClass) ([]NodeImage, error)
+}
+
+func (p *Provider) List(ctx context.Context, nodeClass *v1alpha2.AKSNodeClass) ([]NodeImage, error) {
+	kubernetesVersion, err := p.KubeServerVersion(ctx)
+	if err != nil {
+		return []NodeImage{}, err
+	}
+
+	supportedImages := getSupportedImages(nodeClass.Spec.ImageFamily)
+	useSIG := options.FromContext(ctx).UseSIG
+
+	key, err := p.cacheKey(
+		supportedImages,
+		kubernetesVersion,
+		useSIG,
+	)
+	if err != nil {
+		return []NodeImage{}, err
+	}
+
+	if nodeImages, ok := p.nodeImagesCache.Get(key); ok {
+		return nodeImages.([]NodeImage), nil
+	}
+
+	var nodeImages []NodeImage
+	if useSIG {
+		nodeImages, err = p.listSIG(ctx, supportedImages)
+		if err != nil {
+			return []NodeImage{}, err
+		}
+	} else {
+		nodeImages, err = p.listCIG(ctx, supportedImages)
+		if err != nil {
+			return []NodeImage{}, err
+		}
+	}
+	p.nodeImagesCache.Set(key, nodeImages, imageExpirationInterval)
+
+	return nodeImages, nil
+}
+
+func (p *Provider) listSIG(ctx context.Context, supportedImages []DefaultImageOutput) ([]NodeImage, error) {
+	nodeImages := []NodeImage{}
+	retrievedLatestImages, err := p.NodeImageVersions.List(ctx, p.location, p.subscription)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, supportedImage := range supportedImages {
+		var nextImage *NodeImageVersion
+		for _, retrievedLatestImage := range retrievedLatestImages.Values {
+			if supportedImage.ImageDefinition == retrievedLatestImage.SKU {
+				nextImage = &retrievedLatestImage
+				break
+			}
+		}
+		if nextImage == nil {
+			// Unable to find given image version
+			continue
+		}
+		imageID := fmt.Sprintf(sharedImageGalleryImageIDFormat, options.FromContext(ctx).SIGSubscriptionID, supportedImage.GalleryResourceGroup, supportedImage.GalleryName, supportedImage.ImageDefinition, nextImage.Version)
+
+		nodeImages = append(nodeImages, NodeImage{
+			ID:           imageID,
+			Requirements: supportedImage.Requirements,
+		})
+	}
+	return nodeImages, nil
+}
+
+func (p *Provider) listCIG(_ context.Context, supportedImages []DefaultImageOutput) ([]NodeImage, error) {
+	nodeImages := []NodeImage{}
+	for _, supportedImage := range supportedImages {
+		cigImageID, err := p.getCIGImageID(supportedImage.PublicGalleryURL, supportedImage.ImageDefinition)
+		if err != nil {
+			return nil, err
```

Comment:
Can we consider continuing on error? This returns from within `for` loop. If one image has a problem, others might still work ...

Reply:
We could possibly. However, then we'd have to have some way to still determine if the change in returned images was from a failure, or selectors changing. Maybe we could only use the partially found case, ignoring errors if the NodeImages were empty on the NodeClass. Interesting idea. However, I'm going to leave this as a hard failure for this PR.

Reply:
I also was hoping to refactor this to get a List of the images from one query. However, it looks like how things are setup for the CIGs, that doesn't work within the API